### PR TITLE
Implementation of rotating structures

### PR DESCRIPTION
**Pull request by TwistedAsylumMC**
This pull request adds a method to rotate a structure a given amount of times.

### CODE REVIEW

File: structure.go [hunk 1]
```diff
@@ -106,3 +106,42 @@ func (s Structure) UsePalette(name string) {
 	s.palette = &p
 	s.paletteName = name
 }
+
+// WithRotation returns a new structure with the same contents but rotated 90 degrees clockwise the specified
+// amount of times.
+func (s Structure) WithRotation(iterations int) Structure {
+	iterations = iterations % 4 // Make sure iterations is between 0 and 3.
```
Comment (T14Raptor) [inline structure.go]: I would add constants for this.

File: structure.go [hunk 2]
```diff
@@ -106,3 +106,42 @@ func (s Structure) UsePalette(name string) {
...
+	sizeX, sizeY, sizeZ := int(s.Size[0]), int(s.Size[1]), int(s.Size[2])
+
+	var newStructure Structure
+	switch iterations {
+	case 0, 2:
```
Comment (T14Raptor) [inline structure.go]: I would just return if there's no iterations.

Comment (TwistedAsylumMC) [reply structure.go]: At 180 degrees it has rotations, I guess i'll remove the 0 degrees and return the same structure there instead

**Comment by Sandertv** [comment]
I personally think a `RotateRight90` and `RotateLeft90` method would work. These also already exist for the `cube.Direction` type, so we'd be able to call those respective methods to rotate those blocks.

Another thing I'm considering is that it wouldn't technically be necessary to actually change the underlying structure when rotating. We could simply change the dimensions and apply the transformation (x, z => z, x for example) in the At method. That way you don't actually have to do the extra work. We could then maybe change it only on save?

**Comment by Sandertv** [comment]
Although I suppose if we rotate individual blocks we'd have to actually change the underlying structure too, so nevermind that.

File: structure.go [hunk 3]
```diff
@@ -114,3 +118,81 @@ func (s Structure) UsePalette(name string) {
...
+			fieldV := v.Field(i)
+			if !ast.IsExported(t.Field(i).Name) {
+				continue
+			}
+
+			// We set the field of the original to the new one to make sure that the existing values are placed
+			// back in the form element. Not doing so would result in unexpected behaviour.
+			fieldV.Set(origin.Field(i))
+
+			switch t := fieldV.Interface().(type) {
```
Comment (Sandertv) [inline structure.go]: I'd also add support for `cube.Face` and `cube.Orientation`. Both of those types also have a `RotateRight()` and `RotateLeft()` method.

**Review by Sandertv — COMMENTED** [review]
Got one comment, looks good otherwise.

File: structure.go (hunk 3 above)
Comment (Sandertv) [reply structure.go]: You could consider using `(reflect.Value).MethodByName(...).Call()` to call the `RotateRight()` and `RotateLeft()` methods for all of these types in the same way. All of them have these methods.

**Review by Sandertv — APPROVED** [review]
I think that should do the job! Thanks for the PR!

**Pull request merged**